File: rs/src/basic_data_types.rs
use std::convert::TryInto;
use std::io;

use swf_fixed::{Sfixed16P16, Sfixed8P8};
use swf_tree as ast;

use crate::bit_count::get_i32_min_bit_count;
use crate::io_bits::{BitsWriter, WriteBits};
use crate::primitives::emit_u8;

/// Emits a null-terminated string.
pub fn emit_c_string<W: io::Write>(writer: &mut W, value: &str) -> io::Result<()> {
  writer.write_all(value.as_bytes())?;
  writer.write_all(&[0])
}

pub fn emit_leb128_u32<W: io::Write>(writer: &mut W, mut value: u32) -> io::Result<()> {
  if value == 0 {
    return emit_u8(writer, 0);
  }
  while value != 0 {
    let mut next_byte: u8 = (value & 0x7f).try_into().unwrap();
    value = value >> 7;
    if value != 0 {
      next_byte |= 0x80;
    }
    emit_u8(writer, next_byte)?;
  }
  Ok(())
}

pub fn emit_rect<W: io::Write>(writer: &mut W, value: &ast::Rect) -> io::Result<()> {
  let mut bits_writer = BitsWriter::new(Vec::new());
  emit_rect_bits(&mut bits_writer, value)?;
  writer.write_all(&bits_writer.into_inner()?)
}

pub fn emit_rect_bits<W: WriteBits>(writer: &mut W, value: &ast::Rect) -> io::Result<()> {
  let bits = get_i32_min_bit_count(vec![
    value.x_min,
    value.x_max,
    value.y_min,
    value.y_max,
  ].into_iter());
  writer.write_u32_bits(5, bits)?;
  writer.write_i32_bits(bits, value.x_min)?;
  writer.write_i32_bits(bits, value.x_max)?;
  writer.write_i32_bits(bits, value.y_min)?;
  writer.write_i32_bits(bits, value.y_max)
}

pub fn emit_s_rgb8<W: io::Write + ?Sized>(writer: &mut W, value: ast::SRgb8) -> io::Result<()> {
  emit_u8(writer, value.r)?;
  emit_u8(writer, value.g)?;
  emit_u8(writer, value.b)
}

pub fn emit_straight_s_rgba8<W: io::Write + ?Sized>(writer: &mut W, value: ast::StraightSRgba8) -> io::Result<()> {
  emit_u8(writer, value.r)?;
  emit_u8(writer, value.g)?;
  emit_u8(writer, value.b)?;
  emit_u8(writer, value.a)
}

pub fn emit_matrix<W: io::Write + ?Sized>(writer: &mut W, value: &ast::Matrix) -> io::Result<()> {
  let mut bits_writer = BitsWriter::new(Vec::new());
  emit_matrix_bits(&mut bits_writer, value)?;
  writer.write_all(&bits_writer.into_inner()?)
}

pub fn emit_matrix_bits<W: WriteBits>(writer: &mut W, value: &ast::Matrix) -> io::Result<()> {
  if value.scale_x == Sfixed16P16::ONE && value.scale_y == Sfixed16P16::ONE {
    writer.write_bool_bits(false)?;
  } else {
    writer.write_bool_bits(true)?;
    let bits = get_i32_min_bit_count(vec![value.scale_x.epsilons, value.scale_y.epsilons].into_iter());
    writer.write_u32_bits(5, bits)?;
    writer.write_i32_bits(bits, value.scale_x.epsilons)?;
    writer.write_i32_bits(bits, value.scale_y.epsilons)?;
  }

  if value.rotate_skew0 == Sfixed16P16::ZERO && value.rotate_skew1 == Sfixed16P16::ZERO {
    writer.write_bool_bits(false)?;
  } else {
    writer.write_bool_bits(true)?;
    let bits = get_i32_min_bit_count(vec![value.rotate_skew0.epsilons, value.rotate_skew1.epsilons].into_iter());
    writer.write_u32_bits(5, bits)?;
    writer.write_i32_bits(bits, value.rotate_skew0.epsilons)?;
    writer.write_i32_bits(bits, value.rotate_skew1.epsilons)?;
  }

  {
    let bits = get_i32_min_bit_count(vec![value.translate_x, value.translate_y].into_iter());
    writer.write_u32_bits(5, bits)?;
    writer.write_i32_bits(bits, value.translate_x)?;
    writer.write_i32_bits(bits, value.translate_y)?;
  }

  Ok(())
}

pub fn emit_color_transform<W: io::Write>(writer: &mut W, value: &ast::ColorTransform) -> io::Result<()> {
  let mut bits_writer = BitsWriter::new(Vec::new());
  emit_color_transform_bits(&mut bits_writer, value)?;
  writer.write_all(&bits_writer.into_inner()?)
}

pub fn emit_color_transform_bits<W: WriteBits>(writer: &mut W, value: &ast::ColorTransform) -> io::Result<()> {
  let has_add = value.red_add != 0 || value.green_add != 0 || value.blue_add != 0;
  let has_mult = value.red_mult != Sfixed8P8::ONE || value.green_mult != Sfixed8P8::ONE || value.blue_mult != Sfixed8P8::ONE;

  let mut to_write: Vec<i32> = Vec::new();
  if has_mult {
    to_write.extend_from_slice(&[
      value.red_mult.epsilons.into(),
      value.green_mult.epsilons.into(),
      value.blue_mult.epsilons.into(),
    ]);
  }
  if has_add {
    to_write.extend_from_slice(&[
      value.red_add.into(),
      value.green_add.into(),
      value.blue_add.into(),
    ]);
  }

  let bits = get_i32_min_bit_count(to_write.clone().into_iter());

  writer.write_bool_bits(has_add)?;
  writer.write_bool_bits(has_mult)?;
  writer.write_u32_bits(4, bits)?;

  for value in to_write {
    writer.write_i32_bits(bits, value)?;
  }

  Ok(())
}

pub fn emit_color_transform_with_alpha<W: io::Write>(writer: &mut W, value: &ast::ColorTransformWithAlpha) -> io::Result<()> {
  let mut bits_writer = BitsWriter::new(Vec::new());
  emit_color_transform_with_alpha_bits(&mut bits_writer, value)?;
  writer.write_all(&bits_writer.into_inner()?)
}

pub fn emit_color_transform_with_alpha_bits<W: WriteBits>(writer: &mut W, value: &ast::ColorTransformWithAlpha) -> io::Result<()> {
  let has_add = value.red_add != 0 || value.green_add != 0 || value.blue_add != 0 || value.alpha_add != 0;
  let has_mult = value.red_mult != Sfixed8P8::ONE || value.green_mult != Sfixed8P8::ONE || value.blue_mult != Sfixed8P8::ONE || value.alpha_mult != Sfixed8P8::ONE;

  let mut to_write: Vec<i32> = Vec::new();
  if has_mult {
    to_write.extend_from_slice(&[
      value.red_mult.epsilons.into(),
      value.green_mult.epsilons.into(),
      value.blue_mult.epsilons.into(),
      value.alpha_mult.epsilons.into(),
    ]);
  }
  if has_add {
    to_write.extend_from_slice(&[
      value.red_add.into(),
      value.green_add.into(),
      value.blue_add.into(),
      value.alpha_add.into(),
    ]);
  }

  let bits = get_i32_min_bit_count(to_write.clone().into_iter());

  writer.write_bool_bits(has_add)?;
  writer.write_bool_bits(has_mult)?;
  writer.write_u32_bits(4, bits)?;

  for value in to_write {
    writer.write_i32_bits(bits, value)?;
  }

  Ok(())
}

File: ts/src/lib/emitters/basic-data-types.ts
import { WritableBitStream, WritableByteStream } from "@open-flash/stream";
import { Uint16, UintSize } from "semantic-types";
import { ColorTransform, ColorTransformWithAlpha, Matrix, Rect, SRgb8, StraightSRgba8 } from "swf-types";

import { getSintMinBitCount } from "../get-bit-count.js";

export function emitRect(byteStream: WritableByteStream, value: Rect): void {
  const bitStream: WritableBitStream = byteStream.asBitStream();
  emitRectBits(bitStream, value);
  bitStream.align();
}

export function emitRectBits(bitStream: WritableBitStream, value: Rect): void {
  const nBits: UintSize = getSintMinBitCount(value.xMin, value.xMax, value.yMin, value.yMax);
  bitStream.writeUint16Bits(5, nBits);
  bitStream.writeSint16Bits(nBits, value.xMin);
  bitStream.writeSint16Bits(nBits, value.xMax);
  bitStream.writeSint16Bits(nBits, value.yMin);
  bitStream.writeSint16Bits(nBits, value.yMax);
}

export function emitSRgb8(byteStream: WritableByteStream, value: SRgb8): void {
  byteStream.writeUint8(value.r);
  byteStream.writeUint8(value.g);
  byteStream.writeUint8(value.b);
}

export function emitStraightSRgba8(byteStream: WritableByteStream, value: StraightSRgba8): void {
  byteStream.writeUint8(value.r);
  byteStream.writeUint8(value.g);
  byteStream.writeUint8(value.b);
  byteStream.writeUint8(value.a);
}

export function emitMatrix(byteStream: WritableByteStream, value: Matrix): void {
  const bitStream: WritableBitStream = byteStream.asBitStream();
  emitMatrixBits(bitStream, value);
  bitStream.align();
}

export function emitMatrixBits(bitStream: WritableBitStream, value: Matrix): void {
  if (value.scaleX.valueOf() === 1 && value.scaleY.valueOf() === 1) {
    bitStream.writeBoolBits(false);
  } else {
    bitStream.writeBoolBits(true);
    const nBits: UintSize = getSintMinBitCount(value.scaleX.epsilons, value.scaleY.epsilons);
    bitStream.writeUint16Bits(5, nBits);
    bitStream.writeSint32Bits(nBits, value.scaleX.epsilons);
    bitStream.writeSint32Bits(nBits, value.scaleY.epsilons);
  }
  if (value.rotateSkew0.valueOf() === 0 && value.rotateSkew1.valueOf() === 0) {
    bitStream.writeBoolBits(false);
  } else {
    bitStream.writeBoolBits(true);
    const nBits: UintSize = getSintMinBitCount(value.rotateSkew0.epsilons, value.rotateSkew1.epsilons);
    bitStream.writeUint16Bits(5, nBits);
    bitStream.writeSint32Bits(nBits, value.rotateSkew0.epsilons);
    bitStream.writeSint32Bits(nBits, value.rotateSkew1.epsilons);
  }
  {
    const nBits: UintSize = getSintMinBitCount(value.translateX, value.translateY);
    bitStream.writeUint16Bits(5, nBits);
    bitStream.writeSint16Bits(nBits, value.translateX);
    bitStream.writeSint16Bits(nBits, value.translateY);
  }
}

export function emitColorTransform(byteStream: WritableByteStream, value: ColorTransform): void {
  const bitStream: WritableBitStream = byteStream.asBitStream();
  emitColorTransformBits(bitStream, value);
  bitStream.align();
}

export function emitColorTransformBits(bitStream: WritableBitStream, value: ColorTransform): void {
  const hasAdd: boolean = value.redAdd !== 0 || value.greenAdd !== 0 || value.blueAdd !== 0;
  const hasMult: boolean = value.redMult.valueOf() !== 1
    || value.greenMult.valueOf() !== 1
    || value.blueMult.valueOf() !== 1;

  const toWrite: Uint16[] = [];
  if (hasMult) {
    toWrite.push(value.redMult.epsilons);
    toWrite.push(value.greenMult.epsilons);
    toWrite.push(value.blueMult.epsilons);
  }
  if (hasAdd) {
    toWrite.push(value.redAdd);
    toWrite.push(value.greenAdd);
    toWrite.push(value.blueAdd);
  }

  const bits: UintSize = getSintMinBitCount(...toWrite);

  bitStream.writeBoolBits(hasAdd);
  bitStream.writeBoolBits(hasMult);
  bitStream.writeUint16Bits(4, bits);
  for (const value of toWrite) {
    bitStream.writeSint16Bits(bits, value);
  }
}

export function emitColorTransformWithAlpha(byteStream: WritableByteStream, value: ColorTransformWithAlpha): void {
  const bitStream: WritableBitStream = byteStream.asBitStream();
  emitColorTransformWithAlphaBits(bitStream, value);
  bitStream.align();
}

export function emitColorTransformWithAlphaBits(bitStream: WritableBitStream, value: ColorTransformWithAlpha): void {
  const hasAdd: boolean = value.redAdd !== 0 || value.greenAdd !== 0 || value.blueAdd !== 0 || value.alphaAdd !== 0;
  const hasMult: boolean = value.redMult.valueOf() !== 1
    || value.greenMult.valueOf() !== 1
    || value.blueMult.valueOf() !== 1
    || value.alphaMult.valueOf() !== 1;

  const toWrite: Uint16[] = [];
  if (hasMult) {
    toWrite.push(value.redMult.epsilons);
    toWrite.push(value.greenMult.epsilons);
    toWrite.push(value.blueMult.epsilons);
    toWrite.push(value.alphaMult.epsilons);
  }
  if (hasAdd) {
    toWrite.push(value.redAdd);
    toWrite.push(value.greenAdd);
    toWrite.push(value.blueAdd);
    toWrite.push(value.alphaAdd);
  }

  const bits: UintSize = getSintMinBitCount(...toWrite);

  bitStream.writeBoolBits(hasAdd);
  bitStream.writeBoolBits(hasMult);
  bitStream.writeUint16Bits(4, bits);
  for (const value of toWrite) {
    bitStream.writeSint16Bits(bits, value);
  }
}
